### ISSUE TITLE
runScGen and runDESC contain hardcoded paths

### ISSUE THREAD

**Issue by lazappi** [theislab/scib#158]
In `scib.integration` the `runScGen()` and `runDESC()` function has arguments for temporary directories, e.g. for DESC:

https://github.com/theislab/scib/blob/235faf562bafdfdf3a34b4cad91d6b155cc433f3/scIB/integration.py#L329

But there is no way to set these in the `runIntegration.py` script or the Snakemake `config.yml` (as far as I can tell) which means these methods fail if the paths don't exist.

**Comment by LuckyMD** [comment]
Have you seen these functions fail? I think we tried to avoid saving any models, but have the directories in the arguments as we didn't manage that initially. We didn't want to alter the whole pipeline structure for something we don't actually want to have (model saving).

**Comment by lazappi** [comment]
I have been trying to get the pipeline working on my laptop and had DESC fail because `/localscratch/tmp_desc` didn't exist, that's how I found the problem. I assume it will be an issue for anyone who doesn't have this directory?

Ideally, I guess you would write to the normal tmp directory but I suspect trying to do that also causes issues. Maybe just using wherever the final `adata` is written to would be most portable? You could clear anything that is created before returning from the function if you don't want to leave anything hanging around.

**Comment by LuckyMD** [comment]
Yeah, I vaguely recall turning off writing the model didn't work for DESC. We were initially writing to `/tmp/`, but that turned out not to be appropriate for the local servers.

If we write to wherever the `adata` object is written to, then we have to expose the argument to Snakemake, which means we have to treat the `runDESC` function differently from all others, as these don't require a directory input. I'm a bit hesitant to do this right now, so shortly before revisions. But we do need to fix this in general. 

A simple hack is ofc to just download from github, change the hardcoded directory and work with that. But in general our options are:
1. Write to standard `adata` output
2. Write to tmp directory which will be deleted
3. Write to local tmp directory which is read from the config file
4. Figure out how to turn off writing of the model in DESC

From my perspective 1. is a bad idea, as we don't want to allow methods to re-use trained models that are automatically loaded based on some hash key that identifies if this has been trained before. In that case our memory and speed estimation would be incorrect. 2 is possible, but that would make the runtime longer and also cause false estimation of memory & speed. 3 and 4 are probably the best ideas, and I favour 4. I tried to ask how this can be done, but no response yet from the developer.

**Comment by lazappi** [comment]
The other easy hacky workaround is to just create the required directory (if you have permissions). Not very clean but would work.

**Comment by LuckyMD** [comment]
What's your favourite non-hacky option @lazappi ?

**Comment by lazappi** [comment]
🤷‍♂️ Probably 4, ideally this is something that should be managed by the package itself but that requires the authors to do it. I'm not a big fan of files been written without my control (unless it's done in the proper tmp dir), so probably 3 and 1 are my next preferences. 3 is maybe the most transparent and easiest to clean up? Whatever works I guess?

**Comment by LuckyMD** [comment]
@mumichae any input on how easy 3 is to implement?

**Comment by mumichae** [comment]
3 wouldn't be a problem, we just define a temporary directory specified in the config or within the ROOT directory. I'm assuming that each run would create it's own tmp output and we're not reusing it across settings. In that case I'd actually prefer 2. using snakemake's tmp function. I'm not 100% sure about the timing, but I guess removing the tmp file would not be included in the snakemake benchmark. Only thing that we'd need to know are the exact filenames of the tmp output. I'm checking if we could work it with directories...

**Comment by lazappi** [comment]
There is another hardcoded reference to `/localscratch` in the `scib.pp.readConos()` function https://github.com/theislab/scib/blob/55fc223bf6f4168ef633497c6fe02e294d207fca/scIB/preprocessing.py#L489

Not sure if this can be fixed in the same way or needs to be treated differently but it also causes errors if the directory doesn't exist.

**Comment by LuckyMD** [comment]
This is a different situation, and sth that definitely would require going via the config... We store here and read in again via file. @mumichae would snakemake tmp functionality work for this as well? Or would you go with tmp dir specification in config for this (option 3)?

This would be good to have implemented soon as well...

**Comment by mumichae** [comment]
Fort the hardcoded tmp path in the preprocessing script, would it be enough to use python's [`tempfile`](https://docs.python.org/3/library/tempfile.html)? This might also be a solution for the integration run, especially when users don't want to use the pipeline or specify the temporary directory.

**Comment by LuckyMD** [comment]
Yes, that would work. I've not worked with this before. I assume the name of the tempfile will be stored somewhere and therefore can be read in again?

Not sure if we can use it for the integration run as we just have to supply a folder name to the integration functions.

**Comment by mumichae** [comment]
Not quite sure if I understand, but do CONOS or DESC require to read what they wrote into the tmp  directory later in the pipeline?

**Comment by LuckyMD** [comment]
CONOS, yes.. DESC, no. For CONOS it's the preprocessing that we need to do in python, then save a file, then read into R as a CONOS object. For DESC and scGen it was just that these methods output model files that we can't turn off in the case of DESC.
